File: backend/internal/models/skill.go
package models

import (
	"database/sql/driver"
	"encoding/json"
	"errors"
	"time"
)

// SkillLayer Skill 层级类型
type SkillLayer string

const (
	SkillLayerFoundation SkillLayer = "foundation" // 基础层：最通用的基础知识
	SkillLayerDomain     SkillLayer = "domain"     // 领域层：专业领域知识
	SkillLayerTask       SkillLayer = "task"       // 任务层：特定功能的工作流
)

// SkillSourceType Skill 来源类型
type SkillSourceType string

const (
	SkillSourceManual     SkillSourceType = "manual"      // 手动创建
	SkillSourceModuleAuto SkillSourceType = "module_auto" // Module 自动生成
	SkillSourceHybrid     SkillSourceType = "hybrid"      // 自动生成后手动修改
)

// SkillMetadata Skill 元数据
type SkillMetadata struct {
	Tags        []string `json:"tags,omitempty"`        // 标签（Domain Skill 用于被发现）
	DomainTags  []string `json:"domain_tags,omitempty"` // 需要的领域标签（Task Skill 用于发现 Domain Skills）
	Description string   `json:"description,omitempty"` // 描述
	Author      string   `json:"author,omitempty"`      // 作者
	UsageCount  int      `json:"usage_count,omitempty"` // 使用次数
	AvgRating   float64  `json:"avg_rating,omitempty"`  // 平均评分
}

// Scan 实现 sql.Scanner 接口
func (m *SkillMetadata) Scan(value interface{}) error {
	if value == nil {
		*m = SkillMetadata{}
		return nil
	}

	bytes, ok := value.([]byte)
	if !ok {
		return errors.New("failed to scan SkillMetadata")
	}

	return json.Unmarshal(bytes, m)
}

// Value 实现 driver.Valuer 接口
func (m SkillMetadata) Value() (driver.Value, error) {
	return json.Marshal(m)
}

// Skill AI Skill 知识单元模型
type Skill struct {
	ID             string          `gorm:"primaryKey;type:varchar(36)" json:"id"`
	Name           string          `gorm:"type:varchar(255);not null;uniqueIndex" json:"name"`
	DisplayName    string          `gorm:"type:varchar(255);not null" json:"display_name"`
	Description    string          `gorm:"type:varchar(500)" json:"description"` // 简短描述，用于 AI 智能选择
	Layer          SkillLayer      `gorm:"type:varchar(20);not null" json:"layer"`
	Content        string          `gorm:"type:text;not null" json:"content"`
	Version        string          `gorm:"type:varchar(50);default:'1.0.0'" json:"version"`
	IsActive       bool            `gorm:"default:true" json:"is_active"`
	Priority       int             `gorm:"default:0" json:"priority"`
	SourceType     SkillSourceType `gorm:"type:varchar(50);not null" json:"source_type"`
	SourceModuleID *uint           `gorm:"type:integer" json:"source_module_id,omitempty"`
	Metadata       SkillMetadata   `gorm:"type:jsonb;default:'{}'" json:"metadata"`
	CreatedBy      string          `gorm:"type:varchar(20)" json:"created_by,omitempty"`
	CreatedAt      time.Time       `json:"created_at"`
	UpdatedAt      time.Time       `json:"updated_at"`

	// 关联（非数据库字段）
	SourceModule *Module `gorm:"foreignKey:SourceModuleID" json:"source_module,omitempty"`
}

// TableName 指定表名
func (Skill) TableName() string {
	return "skills"
}

// SkillUsageLog Skill 使用日志模型
type SkillUsageLog struct {
	ID              string    `gorm:"primaryKey;type:varchar(36)" json:"id"`
	SkillIDs        []string  `gorm:"type:jsonb;not null;serializer:json" json:"skill_ids"`
	Capability      string    `gorm:"type:varchar(100);not null" json:"capability"`
	WorkspaceID     string    `gorm:"type:varchar(20)" json:"workspace_id,omitempty"`
	UserID          string    `gorm:"type:varchar(20);not null" json:"user_id"`
	ModuleID        *uint     `gorm:"type:integer" json:"module_id,omitempty"`
	ExecutionTimeMs int       `gorm:"type:integer" json:"execution_time_ms,omitempty"`
	UserFeedback    *int      `gorm:"type:integer" json:"user_feedback,omitempty"`
	AIModel         string    `gorm:"type:varchar(100)" json:"ai_model,omitempty"`
	ContextSummary  string    `gorm:"type:text" json:"context_summary,omitempty"`
	ResponseSummary string    `gorm:"type:text" json:"response_summary,omitempty"`
	CreatedAt       time.Time `json:"created_at"`
}

// TableName 指定表名
func (SkillUsageLog) TableName() string {
	return "skill_usage_logs"
}

// ========== Skill 组合配置结构 ==========

// DomainSkillMode Domain Skill 加载模式
type DomainSkillMode string

const (
	DomainSkillModeFixed  DomainSkillMode = "fixed"  // 固定选择：只使用 domain_skills 中手动选择的
	DomainSkillModeAuto   DomainSkillMode = "auto"   // 自动发现：从 Task Skill 内容中解析依赖
	DomainSkillModeHybrid DomainSkillMode = "hybrid" // 混合模式：固定选择 + 自动发现补充
)

// SkillComposition Skill 组合配置
type SkillComposition struct {
	FoundationSkills    []string               `json:"foundation_skills"`      // Foundation 层 Skill 名称列表
	DomainSkills        []string               `json:"domain_skills"`          // Domain 层 Skill 名称列表（固定选择）
	TaskSkill           string                 `json:"task_skill"`             // Task 层 Skill 名称
	AutoLoadModuleSkill bool                   `json:"auto_load_module_skill"` // 是否自动加载 Module 生成的 Skill
	DomainSkillMode     DomainSkillMode        `json:"domain_skill_mode"`      // Domain Skill 加载模式：fixed/auto/hybrid
	ConditionalRules    []SkillConditionalRule `json:"conditional_rules"`      // 条件加载规则
}

// Scan 实现 sql.Scanner 接口
func (c *SkillComposition) Scan(value interface{}) error {
	if value == nil {
		*c = SkillComposition{}
		return nil
	}

	bytes, ok := value.([]byte)
	if !ok {
		return errors.New("failed to scan SkillComposition")
	}

	return json.Unmarshal(bytes, c)
}

// Value 实现 driver.Valuer 接口
func (c SkillComposition) Value() (driver.Value, error) {
	return json.Marshal(c)
}

// SkillConditionalRule 条件加载规则
type SkillConditionalRule struct {
	Condition string   `json:"condition"`  // 条件表达式，如 "use_cmdb == true"
	AddSkills []string `json:"add_skills"` // 满足条件时额外加载的 Skill 名称列表
}

// ========== 请求/响应结构 ==========

// CreateSkillRequest 创建 Skill 请求
type CreateSkillRequest struct {
	Name        string          `json:"name" binding:"required,max=255"`
	DisplayName string          `json:"display_name" binding:"required,max=255"`
	Description string          `json:"description,omitempty" binding:"max=500"` // 简短描述，用于 AI 智能选择
	Layer       SkillLayer      `json:"layer" binding:"required,oneof=foundation domain task"`
	Content     string          `json:"content" binding:"required"`
	Version     string          `json:"version,omitempty"`
	Priority    int             `json:"priority,omitempty"`
	SourceType  SkillSourceType `json:"source_type,omitempty"`
	Metadata    *SkillMetadata  `json:"metadata,omitempty"`
}

// UpdateSkillRequest 更新 Skill 请求
type UpdateSkillRequest struct {
	DisplayName *string        `json:"display_name,omitempty"`
	Description *string        `json:"description,omitempty"` // 简短描述，用于 AI 智能选择
	Content     *string        `json:"content,omitempty"`
	Version     *string        `json:"version,omitempty"`
	IsActive    *bool          `json:"is_active,omitempty"`
	Priority    *int           `json:"priority,omitempty"`
	Metadata    *SkillMetadata `json:"metadata,omitempty"`
}

// SkillListResponse Skill 列表响应
type SkillListResponse struct {
	Skills     []Skill `json:"skills"`
	Total      int64   `json:"total"`
	Page       int     `json:"page"`
	PageSize   int     `json:"page_size"`
	TotalPages int     `json:"total_pages"`
}

// SkillUsageStats Skill 使用统计
type SkillUsageStats struct {
	SkillID       string     `json:"skill_id"`
	SkillName     string     `json:"skill_name"`
	UsageCount    int64      `json:"usage_count"`
	AvgRating     float64    `json:"avg_rating"`
	AvgExecTimeMs float64    `json:"avg_exec_time_ms"`
	LastUsedAt    *time.Time `json:"last_used_at,omitempty"`
}

// ========== AI 配置扩展 ==========

// AIConfigMode AI 配置模式
type AIConfigMode string

const (
	AIConfigModePrompt AIConfigMode = "prompt" // 提示词模式（默认）
	AIConfigModeSkill  AIConfigMode = "skill"  // Skill 组合模式
)

// ========== 辅助方法 ==========

// IsFoundation 判断是否为 Foundation 层
func (s *Skill) IsFoundation() bool {
	return s.Layer == SkillLayerFoundation
}

// IsDomain 判断是否为 Domain 层
func (s *Skill) IsDomain() bool {
	return s.Layer == SkillLayerDomain
}

// IsTask 判断是否为 Task 层
func (s *Skill) IsTask() bool {
	return s.Layer == SkillLayerTask
}

// IsManual 判断是否为手动创建
func (s *Skill) IsManual() bool {
	return s.SourceType == SkillSourceManual
}

// IsModuleAuto 判断是否为 Module 自动生成
func (s *Skill) IsModuleAuto() bool {
	return s.SourceType == SkillSourceModuleAuto
}

// IsHybrid 判断是否为混合模式（自动生成后手动修改）
func (s *Skill) IsHybrid() bool {
	return s.SourceType == SkillSourceHybrid
}

File: frontend/src/services/skill.ts
import api from './api';

// Skill 层级类型
export type SkillLayer = 'foundation' | 'domain' | 'task';

// Skill 来源类型
export type SkillSourceType = 'manual' | 'module_auto' | 'hybrid';

// Skill 元数据
export interface SkillMetadata {
  tags?: string[];        // 标签（Domain Skill 用于被发现）
  domain_tags?: string[]; // 需要的领域标签（Task Skill 用于发现 Domain Skills）
  description?: string;
  author?: string;
  usage_count?: number;
  avg_rating?: number;
}

// Skill 接口定义
export interface Skill {
  id: string;
  name: string;
  display_name: string;
  description?: string;  // 新增：Skill 描述，用于 AI 智能选择
  layer: SkillLayer;
  content: string;
  version: string;
  is_active: boolean;
  priority: number;
  source_type: SkillSourceType;
  source_module_id?: number;
  metadata?: SkillMetadata;
  created_by?: string;
  created_at: string;
  updated_at: string;
}

// Skill 列表响应
export interface SkillListResponse {
  skills: Skill[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

// 创建 Skill 请求
export interface CreateSkillRequest {
  name: string;
  display_name: string;
  description?: string;  // 新增：Skill 描述
  layer: SkillLayer;
  content: string;
  version?: string;
  priority?: number;
  source_type?: SkillSourceType;
  metadata?: SkillMetadata;
}

// 更新 Skill 请求
export interface UpdateSkillRequest {
  display_name?: string;
  description?: string;  // 新增：Skill 描述
  content?: string;
  version?: string;
  is_active?: boolean;
  priority?: number;
  metadata?: SkillMetadata;
}

// Skill 使用统计
export interface SkillUsageStats {
  skill_id: string;
  skill_name: string;
  usage_count: number;
  avg_rating: number;
  avg_exec_time_ms: number;
  last_used_at?: string;
}

// Domain Skill 加载模式
export type DomainSkillMode = 'fixed' | 'auto' | 'hybrid';

// Domain Skill 模式标签映射
export const DOMAIN_SKILL_MODE_LABELS: Record<DomainSkillMode, string> = {
  fixed: '固定选择',
  auto: '自动发现',
  hybrid: '混合模式',
};

// Domain Skill 模式描述
export const DOMAIN_SKILL_MODE_DESCRIPTIONS: Record<DomainSkillMode, string> = {
  fixed: '只使用手动选择的 Domain Skills（推荐小规模场景）',
  auto: '从 Task Skill 内容中自动发现依赖（推荐大规模场景）',
  hybrid: '固定选择 + 自动发现补充',
};

// Skill 组合配置
export interface SkillComposition {
  foundation_skills: string[];
  domain_skills: string[];
  task_skill: string;
  auto_load_module_skill: boolean;
  domain_skill_mode?: DomainSkillMode; // 新增：Domain Skill 加载模式
  conditional_rules?: ConditionalRule[];
}

// 条件规则
export interface ConditionalRule {
  condition: string;
  add_skills: string[];
}

// 层级标签映射
export const LAYER_LABELS: Record<SkillLayer, string> = {
  foundation: '基础层',
  domain: '领域层',
  task: '任务层',
};

// 来源类型标签映射
export const SOURCE_TYPE_LABELS: Record<SkillSourceType, string> = {
  manual: '手动创建',
  module_auto: 'Module 自动',
  hybrid: '已编辑',
};

// 获取 Skill 列表
export const listSkills = async (params?: {
  layer?: SkillLayer;
  source_type?: SkillSourceType;
  is_active?: boolean;
  search?: string;
  page?: number;
  page_size?: number;
}): Promise<SkillListResponse> => {
  // api 拦截器已经返回 response.data，所以直接返回
  return await api.get('/admin/skills', { params }) as unknown as SkillListResponse;
};

// 获取单个 Skill
export const getSkill = async (id: string): Promise<Skill> => {
  return await api.get(`/admin/skills/${id}`) as unknown as Skill;
};

// 创建 Skill
export const createSkill = async (data: CreateSkillRequest): Promise<Skill> => {
  return await api.post('/admin/skills', data) as unknown as Skill;
};

// 更新 Skill
export const updateSkill = async (id: string, data: UpdateSkillRequest): Promise<Skill> => {
  return await api.put(`/admin/skills/${id}`, data) as unknown as Skill;
};

// 删除 Skill
export const deleteSkill = async (id: string, hard?: boolean): Promise<void> => {
  await api.delete(`/admin/skills/${id}`, { params: { hard } });
};

// 激活 Skill
export const activateSkill = async (id: string): Promise<Skill> => {
  return await api.post(`/admin/skills/${id}/activate`) as unknown as Skill;
};

// 停用 Skill
export const deactivateSkill = async (id: string): Promise<Skill> => {
  return await api.post(`/admin/skills/${id}/deactivate`) as unknown as Skill;
};

// 获取 Skill 使用统计
export const getSkillUsageStats = async (id: string): Promise<SkillUsageStats> => {
  return await api.get(`/admin/skills/${id}/usage-stats`) as unknown as SkillUsageStats;
};

// 获取 Module Skill
export const getModuleSkill = async (moduleId: number): Promise<Skill | null> => {
  try {
    const response = await api.get(`/admin/modules/${moduleId}/skill`) as unknown as { skill: Skill };
    return response.skill;
  } catch (error: any) {
    if (error.includes?.('404') || error.status === 404) {
      return null;
    }
    throw error;
  }
};

// 生成 Module Skill
export const generateModuleSkill = async (moduleId: number): Promise<Skill> => {
  const response = await api.post(`/admin/modules/${moduleId}/skill/generate`) as unknown as { skill: Skill };
  return response.skill;
};

// 更新 Module Skill
export const updateModuleSkill = async (moduleId: number, content: string): Promise<Skill> => {
  const response = await api.put(`/admin/modules/${moduleId}/skill`, { content }) as unknown as { skill: Skill };
  return response.skill;
};

// 预览 Module Skill
export const previewModuleSkill = async (moduleId: number): Promise<string> => {
  const response = await api.get(`/admin/modules/${moduleId}/skill/preview`) as unknown as { preview_content: string };
  return response.preview_content;
};

// ========== Module Version Skill API ==========

export interface ModuleVersionSkill {
  id: string;
  module_version_id: string;
  schema_generated_content: string;
  schema_generated_at?: string;
  schema_version_used?: number;
  custom_content: string;
  inherited_from_version_id?: string;
  combined_content: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

// 获取版本的 Skill
export const getModuleVersionSkill = async (versionId: string): Promise<ModuleVersionSkill> => {
  const response = await api.get(`/admin/module-versions/${versionId}/skill`);
  return response as unknown as ModuleVersionSkill;
};

// 根据 Schema 生成 Skill
export const generateModuleVersionSkill = async (versionId: string): Promise<ModuleVersionSkill> => {
  const response = await api.post(`/admin/module-versions/${versionId}/skill/generate`);
  return response as unknown as ModuleVersionSkill;
};

// 更新自定义内容
export const updateModuleVersionSkillCustomContent = async (
  versionId: string,
  customContent: string
): Promise<ModuleVersionSkill> => {
  const response = await api.put(`/admin/module-versions/${versionId}/skill`, {
    custom_content: customContent,
  });
  return response as unknown as ModuleVersionSkill;
};

// 从其他版本继承 Skill
export const inheritModuleVersionSkill = async (
  versionId: string,
  fromVersionId: string
): Promise<ModuleVersionSkill> => {
  const response = await api.post(`/admin/module-versions/${versionId}/skill/inherit`, {
    from_version_id: fromVersionId,
  });
  return response as unknown as ModuleVersionSkill;
};

// 删除版本的 Skill
export const deleteModuleVersionSkill = async (versionId: string): Promise<void> => {
  await api.delete(`/admin/module-versions/${versionId}/skill`);
};

// ========== Domain Skill 自动发现预览 API ==========

export interface DiscoveredSkillSummary {
  name: string;
  display_name: string;
  tags: string[];
  priority: number;
}

export interface PreviewDiscoveryResponse {
  task_skill: string;
  domain_tags: string[];
  discovered_skills: DiscoveredSkillSummary[];
  discovered_count: number;
  message?: string;
}

// 预览 Domain Skill 自动发现结果
export const previewDomainSkillDiscovery = async (taskSkillName: string): Promise<PreviewDiscoveryResponse> => {
  return await api.get('/admin/skills/preview-discovery', {
    params: { task_skill: taskSkillName }
  }) as unknown as PreviewDiscoveryResponse;
};
